book/src/04_4_0_interrupts.md

```diff
@@ -1,12 +1,20 @@
 # Interrupts
 
-The goal of this exercise is to handle the interrupt that fires if the `BOOT` button is pushed. 
-This exercise involves working with C bindings to the [esp-idf-sys](https://esp-rs.github.io/esp-idf-sys/esp_idf_sys/index.html) and other unsafe operations, as well as non-typical rust documentation. In a first step we will go line by line to build this interrupt handler. 
+An interrupt is a request for the processor to interrupt currently executing code, so that the event can be processed in a timely manner. If the request is accepted, the processor will suspend its current activities, save its state, and execute a function called an interrupt handler to deal with the event. Interrupts are commonly used by hardware devices to indicate electronic or physical state changes that require time-sensitive attention, for example pushing a button. 
 
-You can find a skeleton code for this exercise in `advanced/button-interrupt/exercise/src/main.rs.`
-You can find the solution for this exercise in `advanced/button-interrupt/solution/src/main.rs`
+## Challenges
 
-## A note on `unsafe {}` blocks:
+The fact that interrupt handlers can be called at any time provides a challenge in embedded Rust as this requires the existence of statically allocated mutable memory that both the interrupt handler and the main code can refer to. In Rust such memory can be declared by defining a `static mut`. But reading and writing to such variables is always unsafe, as without precautions race conditions can be triggered. 
+
+How do we handle this problem?
+
+In our example the `static mut` holds the state of the event queue: was there an interrupt or was there no interrupt? We read this state in `main()` only. While this technically can trigger a race condition because a `write` from the interrupt handler can overlap the `read` from `main()`, this would in the worst case cause an unnoticed interrupt. This is unproblematic in a way as other ways to handle this memory safe for example with critical sections or atomic operations would cause the same problem: If they are in action an interrupt would go unnoticed. 
+
+Read more about this in the [Embedded Rust Book](https://docs.rust-embedded.org/book/concurrency/index.html)
```

Comment:
You may also want to note that the interrupt routine has to be in RAM because an interrupt can occur when the Flash Memory is out of action (e.g. during a write to a Flash Filesystem).